### PR TITLE
Unstated 3.0 RFC

### PR DESCRIPTION
**Pull request by jamiebuilds**
### [SEE RENDERED](https://github.com/jamiebuilds/unstated/blob/3.0-rfc/UNSTATED_3_RFC.md)

### CODE REVIEW

File: UNSTATED_3_RFC.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  let [count, setCount] = useState(initialCount);
+  let reset = () => setCount(0);
+  let increment = () => setCount(count => count - amount);
+  let decrement = () => setCount(count => count + amount);
+  return { count, reset, increment, decrement };
+}
+
+function Counter() {
+  // use existing instance from context, or create new one
+  let counter = useProvided(useCounter);
```
Comment (mohammedzamakhan) [inline UNSTATED_3_RFC.md]: for some reason `useProvided` doesn't looks like a good name, may be like `useGlobalState` / `useStateFrom`/ `useInstance`

File: UNSTATED_3_RFC.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    Count: {counter.count}
+    <button onClick={counter.reset}>Reset</button>
+    <button onClick={counter.increment}>+</button>
+    <button onClick={counter.decrement}>-</button>
+  </>;
+}
+
+function App() {
+  // create new instance of container
+  let counter = useInject(useCounter, 42);
```
Comment (vanthiyathevan) [inline UNSTATED_3_RFC.md]: I really liked the auto inject in current. Just need 1 `Provider` at the root and forget. Would this mean you have to inject all context before using them?

Comment (jamiebuilds) [reply UNSTATED_3_RFC.md]: No... it's exactly the same

**Comment by ezekielchentnik** [comment]
can you still use containers outside of the react tree?

**Comment by slorber** [comment]
didn't dig deep in the impl but that is probably a good idea for simple containers

However, in my case I use unstated as a global state container, like a replacement for redux. I absolutely need:

1) Ability to call methods from other methods. Ie define some helpers that do `this.setState` and reuse these methods at some other places. Not sure yet about this but I think the code could look simpler by being class based versus defining many functions in a hook, calling each others... need to think more about this

2) Need setState to return a promise (I guess it could be re-built on top of useState in userland?)

3) Need ability to connect my global state in multiple components in a performant way (ie use a hoc and selectors and make the component pure by default, like redux). As of today there is no hook afaik that permits to replace Redux connect in a performant way, you have to re-render on every global state change all the connected components. But your solution does not prevent me to implement a connect HOC on top of it. The HOC could use hooks to retrieve the data and inject only the selected data slice to the wrapped component.

I'll send you from twitter a private gist if you are interested to understand better my unstated usage.

In any case I think it'll be best to keep both versions around. I'll try to tell you asap if I'm able to migrate to hooks and if the code still looks clean

File: UNSTATED_3_RFC.md (hunk 1 above)
Comment (yesmeck) [reply UNSTATED_3_RFC.md]: How can we create an instance of a hook?

**Comment by AjaxSolutions** [comment]
The increment and decrement should be reversed.  Currently, increment is decreasing the count and increment is decreasing the count.  This looks wrong.
```
function useCounter(initialCount = 0) {
  let [count, setCount] = useState(initialCount);
  let reset = () => setCount(0);
  let increment = () => setCount(count => count - amount);
  let decrement = () => setCount(count => count + amount);
  return { count, reset, increment, decrement };
}
```

**Comment by rbalicki2** [comment]
I'd like to 👍  what @slorber said. In particular, we use `unstated` as follows:

```
// modalContainer.js
class ModalContainer extends Container {
  state = { open: false };
  open = () => this.setState({ open: true });
};
const modalContainer = new ModalContainer();
export default modalContainer;
export const ModalSubscriber = ({ children }) => <Provider><Subscribe to={[modalContainer]}>{ children }</Subscribe></Provider>

// modalButton.js
<ModalSubscriber>{ ({ open}) => <button onClick={open}>Open a modal</button>

// modal.js
<ModalSubscriber>{ ({ state }) => state.open && <ModalComponent /> }</ModalSubscriber>
```

I'm concerned that this new RFC prevents this use case, because hooks only work during a render method. However, we're perfectly happy with the current implementation and would gladly stay on v2.

**Comment by ehahn9** [comment]
currently using (& loving!) unstated. Thanks!!

re 3.0: in the RFC you say:

`// use existing instance from context, or create new one if it doesn't exist`

...but doesn't that mean the `useState` at the top of the custom `useCounter` hook will be called *conditionally* (called to create the shared context, not called otherwise). Isn't conditionally calling hooks a [no-no](https://reactjs.org/docs/hooks-rules.html)?

I really like the simplicity of your 3.0 thinking, but wondering how you get around this difficult design challenge?

**Comment by rohmanhm** [comment]
Hope on Unstated 3.0 has better support for typescript users.

**Comment by charlie632** [comment]
Is this still open? Is there any support for hooks yet?

**Comment by slorber** [comment]
Didn't try but this lib seems like a nice lightweight hooks-based state manager: https://github.com/diegohaz/constate

**Comment by charlie632** [comment]
@slorber thanks! I'll give it a try.

**Comment by alex-pex** [comment]
Maybe you should take a look on the work done in https://github.com/yamalight/outstated#guide
> Oustated is like unstated but with hooks

**Comment by jamiebuilds** [comment]
This is the next version of Unstated: https://github.com/jamiebuilds/unstated-next

**Pull request closed without merging**